### PR TITLE
EE: Remote Repositories

### PR DESCRIPTION
**Pull request by himdel**
Fixes [AAH-437](https://issues.redhat.com/browse/AAH-437) (with #873)
API: https://github.com/ansible/galaxy_ng/pull/935 (merged), https://github.com/ansible/galaxy_ng/pull/985 (merged), https://github.com/ansible/galaxy_ng/pull/988
Design https://marvelapp.com/prototype/27ee1cd6/screen/82150243 
Soft-depends on #873 for Remote Registry crud (merged)

This allows adding, editing and syncing remote repositories in Execution Environments. These new remote container repos act like local ones, except only local ones can have tags managed while only remote ones can sync, and show sync status.

* EE list:
  * [x] Add button to sync EE if it is a remote EE.
  * [x] Add option to edit an EE’s remote.
  * [x] Add button to create new remote EE.
  * [x] Add if container registry is remote or not

* EE detail:
  * [x] last updated from registry
  * [x] edit merge
  * [x] disallow editing tags on remotes
  * [x] dropdown with sync&edit(&use)

* the form itself:
  * [x] Create new remote repository: Allows the user to specify an upstream image name, registry and image name. Creates a new remote execution environment that can be synced
  * [x] Edit remote form. Allows updating a remotes upstream name and registry BUT NOT THE REMOTE’S NAME.
  * [x] rename from Edit repository to Edit execution environment

|  |  |
|-|-|

### CODE REVIEW

**Comment by himdel** [comment]
(Currently this allows Add/Edit of remote repos from the EE list screen, except for groups&description, so I mostly just need to merge the 2 `onSave` bits to do both in both places)

**Comment by himdel** [comment]
Testing:

registry url: `https://registry.hub.docker.com`
remote upstream_name: `library/alpine` (`_/alpine` on dockerhub becomes `library/alpine`)

**Review by ZitaNemeckova — APPROVED** [review]
LGTM :+1:

**Comment by newswangerd** [comment]
Need to add the "create ee" button to the empty state page when there are no execution enviromments. Right now it's impossible to add a remote ee if the registry is empty.
Screen Shot 2021-09-24 at 3 07 34 PM

File: src/containers/execution-environment-list/execution_environment_list.tsx [hunk 1]
```diff
@@ -338,13 +386,68 @@ class ExecutionEnvironmentList extends React.Component<
...
         </td>
       </tr>
     );
   }
 
+  private renderRemoteModal(itemToEdit) {
+    const { name, namespace, description, pulp } = itemToEdit;
+    const { pulp_id, registry, upstream_name, include_tags, exclude_tags } =
+      pulp?.repository?.remote || {};
+    const remote = pulp?.repository ? !!pulp?.repository?.remote : true; // add only supports remote
```
Comment (newswangerd) [inline src/containers/execution-environment-list/execution_environment_list.tsx:403]: Don't use `object?.` It's not supported on older browsers and we don't have a polyfill.

File: src/components/execution-environment/repository-form.tsx [hunk 2]
```diff
@@ -114,4 +340,120 @@ export class RepositoryForm extends React.Component<IProps, IState> {
...
+  private removeTag(tag, key: 'includeTags' | 'excludeTags') {
+    const current = new Set(this.state[key]);
+    current.delete(tag);
+
+    this.setState({
+      [key]: Array.from(current.values()),
+    } as any);
+  }
+
+  private onSave() {
```
Comment (newswangerd) [inline src/components/execution-environment/repository-form.tsx:383]: I don't know where this is happening, but if the tasking system is busy (if a sync is happening) and you try to edit the description on an repository, the page spams the task endpoint for updates on the distribution task. If you navigate away from the page while the form is loading, the UI seems to keep spamming the API forever.

You need to rate limit the requests for updates from the task to 1 every 5 seconds or so and make sure to garbage collect the request if the component is unmounted.

**Review by newswangerd — CHANGES_REQUESTED** [review]
Getting an error on the images page after performing a sync

```
Uncaught TypeError: selectedImage is undefined
    renderImages execution_environment_detail_images.tsx:186
    render execution_environment_detail_images.tsx:114
    React 11
    unstable_runWithPriority scheduler.development.js:653
    React 6
    queryImages execution_environment_detail_images.tsx:436
```

File: src/containers/execution-environment-list/execution_environment_list.tsx (hunk 1 above)
Comment (himdel) [reply src/containers/execution-environment-list/execution_environment_list.tsx:409]: Looks like `?.` gets compiled by babel, `Array.prototype.at` was the problematic one :)

File: src/components/execution-environment/repository-form.tsx (hunk 2 above)
Comment (himdel) [reply src/components/execution-environment/repository-form.tsx:385]: Updated to use a shared `waitForTask` helper & set the limit to 5s.
(still looking into making sure it ends at some point)

File: src/containers/execution-environment-list/execution_environment_list.tsx (hunk 1 above)
Comment (newswangerd) [reply src/containers/execution-environment-list/execution_environment_list.tsx:409]: my bad :)

**Comment by himdel** [comment]
Updated...

`waitForTask` is now shared, and bails after 10 attempts
`selectedImage` being undefined is ok after deletion, handling the error
and added the Add EE button to empty state :)

**Review by newswangerd — APPROVED** [review]

**Pull request merged**